File: src/puppy_worlds/scripts/clock_publisher.py
#!/usr/bin/env python3
"""Manual /clock publisher for Gazebo simulation.

Reads timestamps from /odom topic (published by Gazebo with sim time)
and republishes them to /clock so nodes with use_sim_time=true can
synchronize properly.

This is needed because gazebo_ros in this environment doesn't publish
/clock automatically.
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy, QoSDurabilityPolicy
from nav_msgs.msg import Odometry
from rosgraph_msgs.msg import Clock
from builtin_interfaces.msg import Time


class ClockPublisher(Node):
    def __init__(self):
        super().__init__('clock_publisher')
        
        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.VOLATILE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=10,
        )
        
        self.clock_pub = self.create_publisher(
            Clock, '/clock', qos_profile)
        
        self.odom_sub = self.create_subscription(
            Odometry, '/odom', self.odom_callback, 10)
        
        self.last_time = None
        self.get_logger().info('Clock publisher started - reading from /odom')
    
    def odom_callback(self, msg):
        """Extract sim time from odometry message and publish to /clock."""
        clock_msg = Clock()
        clock_msg.clock = msg.header.stamp
        self.clock_pub.publish(clock_msg)
        self.last_time = msg.header.stamp


def main(args=None):
    rclpy.init(args=args)
    node = ClockPublisher()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp_src/puppy_worlds/clock_publisher.cpp
// Manual /clock publisher 实现
//
// 从 /odom 提取仿真时间并发布到 /clock
#include "puppy_worlds/clock_publisher.h"

namespace puppy_worlds {

ClockPublisher::ClockPublisher() : rclcpp::Node("clock_publisher") {
  // 等价 Python:
  //   qos_profile = QoSProfile(
  //       reliability=QoSReliabilityPolicy.BEST_EFFORT,
  //       durability=QoSDurabilityPolicy.VOLATILE,
  //       history=QoSHistoryPolicy.KEEP_LAST,
  //       depth=10,
  //   )
  rclcpp::QoS qos_profile(10);  // KEEP_LAST, depth=10
  qos_profile.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);
  qos_profile.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);

  // /clock 发布器
  clock_pub_ = this->create_publisher<rosgraph_msgs::msg::Clock>("/clock",
                                                                 qos_profile);

  // /odom 订阅器
  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "/odom", 10,
      std::bind(&ClockPublisher::odomCallback, this, std::placeholders::_1));

  has_last_time_ = false;
  RCLCPP_INFO(this->get_logger(), "Clock publisher started - reading from /odom");
}

void ClockPublisher::odomCallback(
    const nav_msgs::msg::Odometry::SharedPtr msg) {
  // 从里程计消息中提取仿真时间并发布到 /clock
  // 等价 Python:
  //   clock_msg = Clock()
  //   clock_msg.clock = msg.header.stamp
  //   self.clock_pub.publish(clock_msg)
  //   self.last_time = msg.header.stamp
  rosgraph_msgs::msg::Clock clock_msg;
  clock_msg.clock = msg->header.stamp;
  clock_pub_->publish(clock_msg);
  last_time_ = msg->header.stamp;
  has_last_time_ = true;
}

}  // namespace puppy_worlds

// ===== 节点入口 =====
int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<puppy_worlds::ClockPublisher>());
  rclcpp::shutdown();
  return 0;
}
